hal/src/thumbv7em/clock/v2.rs

```diff
@@ -1,98 +1,859 @@
+//! [*identity*](self#id-types) of the XOSC0 clock, rather than the clock
+//! itself, and `M` represents the XOSC [`Mode`](xosc::Mode).
+//!
+//! Next, we access the [`Tokens`] struct to extract the corresponding
+//! [`XoscToken`] for XOSC0, and we trade the PAC `PORT` struct for the
+//! [`gpio::Pins`] struct to access the GPIO pins. We can then call
+//! [`Xosc::from_crystal`] to trade the token and [`Pin`]s to yield an instance
+//! of [`Xosc0`]. In doing so, we also provide the oscillator frequency.
+//!
+//! Finally, we can chain a call to the [`Xosc::enable`] method to enable the
+//! XOSC and return an instance of [`EnabledXosc0<M, N>`], which is simply an
+//! alias for `Enabled<Xosc0<M>, N>`. In this case, we get
+//! `EnabledXosc0<CrystalMode, U0>`.
+//!
+//! ```no_run
+//! # use atsamd_hal::{
+//! #     clock::v2::{
+//! #         clock_system_at_reset,
+//! #         xosc::Xosc,
+//! #     },
+//! #     gpio::Pins,
+//! #     pac::Peripherals,
+//! #     time::U32Ext,
+//! # };
+//! # let mut pac = Peripherals::take().unwrap();
+//! # let (buses, clocks, tokens) = clock_system_at_reset(
+//! #     pac.OSCCTRL,
+//! #     pac.OSC32KCTRL,
+//! #     pac.GCLK,
+//! #     pac.MCLK,
+//! #     &mut pac.NVMCTRL,
+//! # );
+//! let pins = Pins::new(pac.PORT);
+//! let xosc0 = Xosc::from_crystal(
+//!     tokens.xosc0,
+//!     pins.pa14,
+//!     pins.pa15,
+//!     8.mhz(),
+//! ).enable();
+//! ```
+//!
+//! Next, we want to use a DPLL to multiply the 8 MHz crystal clock up to 100
+//! MHz. Once again, we need to decide between two instances of a clock, because
+//! each chip has two [`Dpll`]s. This time, however, our decision between
+//! [`Dpll0`] and [`Dpll1`] is arbitrary.
+//!
+//! Also note that, like before, `Dpll0<I>` and `Dpll1<I>` are aliases for
+//! `Dpll<Dpll0Id, I>` and `Dpll<Dpll1Id, I>`. [`Dpll0Id`] and [`Dpll1Id`]
+//! represent the *identity* of the respective DPLL, while `I` represents the
+//! [`Id` type](self#id-types) for the [`Source`] driving the DPLL. In this
+//! particular case, we aim to create an instance of `Dpll0<Xosc0Id>`.
 //!
-//! This module provides the set of abstractions allowing a user to manipulate
-//! an ATSAMD's clocking system in a safe and expressive manner.
+//! Only certain clocks can drive the DPLL, so `I` is constrained by the
+//! [`DpllSourceId`] trait. Specifically, only the [`Xosc0Id`], [`Xosc1Id`],
+//! [`Xosc32kId`] and [`GclkId`] types implement this trait.
 //!
-//! Foundation of an API is a [`retrieve_clocks`] function that returns a tuple
-//! of instantiated clocking abstractions that are enabled on reset.
+//! As before, we access the [`Tokens`] struct and use the corresponding
+//! [`DpllToken`] when creating an instance of `Dpll`. However, unlike before,
+//! we are creating a new clock-tree relationship that must be tracked by the
+//! type system. Because DPLL0 will now consume XOSC0, we must [`Increment`]
+//! the [`Enabled`] [`Counter`] for [`EnabledXosc0`].
```

Comment:
`Counter` is `Enabled`? Hmm :/

Reply:
I'm not sure about `we must` phrasing used together with `Increment` and `Decrement` later on. Does it suggest that user should decrement/increment some counters themselves or is it obvious that `from_source` and `free_source` do that themselves?

Reply:
> Because DPLL0 will now consume XOSC0, the `Counter` in `EnabledXosc0` is going to get incremented.

Now when I think about it, `consume` can be misleading for some ppl thinking about Rust's `consuming` of a variable? Hmm. I might be overreacting though.